### DISCUSSION TITLE
Golang plugin maintain singleton db connection pool

### DISCUSSION THREAD

**Discussion by dashng** [Kong/kong#9767] [Help]
I am working on a kong golang plugin. 
The plugin is to save the user request logs. I face a issue that the server running slower after the server started for hours.
I checked the log and found the db connection pool is initialized many times.
How can I singleton a db connection pool during the whole app lifecycle.
Appreciate if anyone can help on this!!

**Comment by StarlightIbuki** [comment]
Could you share more information about this issue? Theoretically, GO-PDK should not access the database.

**Comment by dashng** [comment]
We are trying to initialize the db connection only once when the plugin is bootstraped.

But it seems the once function is not working properly, the data base connection init function is called every time when the new request coming.

Do we have a way to only run function once except for putting the code to init func.

```
var once sync.Once
func (conf Config) Access(kong *pdk.PDK) {
	fun1 := func() {
		DB_HOST := getenv("DB_HOST", "kong")
		DB_PORT := getenv("DB_PORT", "5432")
		DB_DBNAME := getenv("DB_DBNAME", "dbname")
		DB_USER := getenv("DB_USER", "postgres")
		DB_PASSWORD := getenv("DB_PASSWORD", "postgres")
		DB_SSLMODE := getenv("DB_SSLMODE", "disable")
		DB_TIMEZONE := getenv("DB_TIMEZONE", "Asia/Shanghai")
		REDIS_HOST := getenv("REDIS_HOST", "redis")
		REDIS_PORT := getenv("REDIS_PORT", "6380")
		REDIS_PWD := getenv("REDIS_PASSWORD", "redis123!")

		initServices(
			DB_HOST,
			DB_PORT,
			DB_DBNAME,
			DB_USER,
			DB_PASSWORD,
			DB_SSLMODE,
			DB_TIMEZONE,
			REDIS_HOST,
			REDIS_PORT,
			REDIS_PWD,
		)
	}
	once.Do(fun1)
}
```

**Reply by dashng** [reply]
Thanks for the explanation. 
Do we have the any best practise for init / maintain db connection?
@StarlightIbuki

**Reply by StarlightIbuki** [reply]
> Thanks for the explanation. Do we have the any best practise for init / maintain db connection? @StarlightIbuki

Sorry, I've confirmed that `go-pdk` is using coroutine, not processes.

I failed to reproduce your problem. The code I'm using:

```go
package main

import (
	"sync"

	"github.com/Kong/go-pdk"
	"github.com/Kong/go-pdk/server"
)

const Version = "1.0"
const Priority = 1

func main() {
	server.StartServer(New, Version, Priority)
}

type Config struct {
	Num string
}

func New() interface{} {
	return &Config{}
}

var once sync.Once

func (conf Config) Access(kong *pdk.PDK) {
	fun1 := func() {
		print("hello ")
	}
	print("hey ")
	once.Do(fun1)
}

```

**Comment by dashng** [comment]
@StarlightIbuki
